ray-operator/test/e2eautoscaler/raycluster_autoscaler_test.go

```diff
@@ -194,3 +196,84 @@ func TestRayClusterAutoscalerWithCustomResource(t *testing.T) {
+
+	var rayCluster *rayv1.RayCluster
+	const maxReplica = 3
+
+	// Used in main goroutine to block wait until kubernetes setup completion.
+	var setupReadyWg sync.WaitGroup
+	// Used in main goroutine to block wait until test script execution completion.
+	var taskExecWg sync.WaitGroup
+
+	// Create a namespace
+	namespace := test.NewTestNamespace()
+	test.StreamKubeRayOperatorLogs()
+
+	// Scripts for creating and terminating detached actors to trigger autoscaling
+	scriptsAC := newConfigMap(namespace.Name, files(test, "create_concurrent_tasks.py"))
+	scripts, err := test.Client().Core().CoreV1().ConfigMaps(namespace.Name).Apply(test.Ctx(), scriptsAC, TestApplyOptions)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	test.T().Logf("Created ConfigMap %s/%s successfully", scripts.Namespace, scripts.Name)
+
+	setupReadyWg.Add(1)
+	taskExecWg.Add(1)
+	go func() {
+		test.T().Run("Create a RayCluster with autoscaling enabled", func(_ *testing.T) {
+			groupName := "custom-resource-group"
+
+			rayClusterSpecAC := rayv1ac.RayClusterSpec().
+				WithEnableInTreeAutoscaling(true).
+				WithRayVersion(GetRayVersion()).
+				WithHeadGroupSpec(rayv1ac.HeadGroupSpec().
+					WithRayStartParams(map[string]string{"num-cpus": "0"}).
+					WithTemplate(headPodTemplateApplyConfiguration())).
+				WithWorkerGroupSpecs(rayv1ac.WorkerGroupSpec().
+					WithReplicas(0).
+					WithMinReplicas(0).
+					WithMaxReplicas(maxReplica).
+					WithGroupName(groupName).
+					WithRayStartParams(map[string]string{"num-cpus": "1", "resources": `'{"CustomResource": 1}'`}).
+					WithTemplate(workerPodTemplateApplyConfiguration()))
+			rayClusterAC := rayv1ac.RayCluster("ray-cluster", namespace.Name).
+				WithSpec(apply(rayClusterSpecAC, mountConfigMap[rayv1ac.RayClusterSpecApplyConfiguration](scripts, "/home/ray/test_scripts")))
+
+			var err error
+			rayCluster, err = test.Client().Ray().RayV1().RayClusters(namespace.Name).Apply(test.Ctx(), rayClusterAC, TestApplyOptions)
+			g.Expect(err).NotTo(gomega.HaveOccurred())
+			test.T().Logf("Created RayCluster %s/%s successfully", rayCluster.Namespace, rayCluster.Name)
+			setupReadyWg.Done()
+
+			// Wait for RayCluster to become ready and verify the number of available worker replicas.
+			g.Eventually(RayCluster(test, rayCluster.Namespace, rayCluster.Name), TestTimeoutMedium).
+				Should(gomega.WithTransform(RayClusterState, gomega.Equal(rayv1.Ready)))
+			g.Expect(GetRayCluster(test, rayCluster.Namespace, rayCluster.Name)).To(gomega.WithTransform(RayClusterDesiredWorkerReplicas, gomega.Equal(int32(0))))
+
+			headPod, err := GetHeadPod(test, rayCluster)
+			g.Expect(err).NotTo(gomega.HaveOccurred())
+			test.T().Logf("Found head pod %s/%s", headPod.Namespace, headPod.Name)
+
+			// Create a number of tasks and wait for their completion, and a worker in the "custom-resource-group" should be created.
+			ExecPodCmd(test, headPod, common.RayHeadContainer, []string{"python", "/home/ray/test_scripts/create_concurrent_tasks.py"})
```

Comment:
I think we don't need `setupReadyWg` and `taskExecWg`. Actually, only this line needs to be executed in parallel. Wrapping this line in a goroutine is enough.

Reply:
For `setupReadyWg`, before goroutine starts, `rayCluster` could be null

Reply:
For `taskExecWg`, you need a way to tell the python script has completed its execution for synchronzation

Reply:
Okay, then only 1 wg is needed if only `ExecPodCmd` is wrapped in goroutine.

Reply:
If line 245 is extracted outside a goroutine, then `rayCluster` will not be null.

Reply:
Ok, updated.